Add a note for contributors installing fourmolu (#10473)
haskell/cabal: CONTRIBUTING.md

## Code Style

We use automated formatting, checked in the CI process, to enforce a unified
style across the code bases and include these makefile targets to help with
formatting:

* `make style` - Format the `Cabal`, `Cabal-syntax` and `cabal-install` directories.
* `make style-modified` - Format files modified in the current tree.
* `make style-commit COMMIT=<ref>` - Format files modified between HEAD and the given reference.

> [!NOTE]
> We use `fourmolu-0.12.0.0` for formatting. If installing it with `cabal

> > install`, please make sure to use a version of `GHC >= 9.2.1 && < 9.8`. That

Reviewer: ```suggestion
> install`, please make sure to use a version of `GHC >= 9.2.1 && < 9.8`.
```

> > it requires `GHC2021` sets the lower bound and its reliance on

Reviewer: ```suggestion
> Its requirement of `GHC2021` sets the lower bound, and its reliance on
```

> > `ghc-lib-parser` sets the upper bound on GHC versions. The command for

Reviewer: ```suggestion
> `ghc-lib-parser` sets the upper bound on GHC versions. A command for
```

> installing it this way is:
>
> ```

> > $ cabal install fourmolu-0.12.0.0 --overwrite-policy=always --ignore-project

Reviewer: ```suggestion
> $ cabal install fourmolu-0.12.0.0 --overwrite-policy=always --ignore-project --with-compiler=ghc-9.6
```

Reviewer 2: Mind you not all distributions create the symlinks like `ghc-9.6`. E.g. `nixpkgs` doesn't (it does create things like `ghc-9.6.5`). So, I'm not big fan of hardcoding these things. There was times when this `with-compiler: 8.x` (or `9.x`)  was hardcoded in cabal's own project file and made contributing to it pretty painful.

> ```

> 

Reviewer: ```suggestion

(This assumes you have a ghc version in the 9.6 series installed via ghcup.)
```

Reviewer 2: ghcup is not available everywhere, e.g. NixOS, so I'd not overfit for it.

Reviewer: I provided it as an example, and it does provide that symlink. I don't think we can provide examples for every possible distribution, but I also don't think we should provide none, so I provided the most common.

Reviewer: In fact, nix would require starting with an appropriate `nix-shell` invocation AIUI, so we'd have trouble providing that example anyway.

Reviewer 2: Not necessarily: `ghc` can be installed in `systemPackaes` on a NixOS system.

Reviewer: But it may be the wrong version, especially since you need an older one for fourmolu 0.12.

I am getting the strong impression that you think hanging new contributors out to dry is better than providing _any_ example. _Really?_

Reviewer: Also, the specific reason for `--with-compiler` is that the original says nothing whatsoever about making sure an appropriate compiler version is in scope for the `cabal install`.

Author: > the original says nothing whatsoever about making sure an appropriate compiler version is in scope

Yes, I went for an incremental improvement and mentioned that the GHC version will need to be restricted. I didn't go so far as to say how to go about that but am happy to. I use ghcup for that (and sometimes nix).

Reviewer 2: I don't suggest removing the example. I only suggest providing essential details and leaving off inessential or potentially problematic ones. For example, the way people obtain GHC is inessential (also, hardcoding a particular distribution channel is not very future-proof). The `with-compiler` is completely optional, as long as the text clearly states a particular version of GHC that should be in scope.

I don't want to drag this thread for too long, so if you feel strongly about `with-compiler` and mentioning ghcup, please go ahead.

Author: Looking at the package bounds on hackage, `fourmolu-0.12.0.0` requires `ghc-lib-parser (>=9.6 && <9.7)` but that doesn't restrict GHC to just the `9.6` series. The constraints on `base` for [ghc-lib-parser-9.6.6.20240701](https://hackage.haskell.org/package/ghc-lib-parser-9.6.6.20240701) are `base (>=4.16.1 && <4.19)`, so from [snoyman/base](https://www.snoyman.com/base/) that is `GHC >= 9.2.2 && GHC < 9.8`.

## Code Style

We use automated formatting, checked in the CI process, to enforce a unified
style across the code bases and include these makefile targets to help with
formatting:

* `make style` - Format the `Cabal`, `Cabal-syntax` and `cabal-install` directories.
* `make style-modified` - Format files modified in the current tree.
* `make style-commit COMMIT=<ref>` - Format files modified between HEAD and the given reference.

> [!NOTE]
> We use `fourmolu-0.12.0.0` for formatting. If installing it with `cabal
> install`, please make sure to use a version of `GHC >= 9.2.1 && < 9.8`.
> Its requirement of `GHC2021` sets the lower bound, and its reliance on
> `ghc-lib-parser` sets the upper bound on GHC versions. A command for
> installing it this way is:
>
> ```

> > $ cabal install fourmolu-0.12.0.0 --overwrite-policy=always --ignore-project

Reviewer: A different worry here is that people may have other versions of fourmolu installed, and may need them for other projects. Sadly, there is no good solution to this at present.

Author: With already having installed `fourmolu` once before, reinstalling is much quicker, isn't it? For me, this seems to be almost instant and, I would guess, only involves resetting the symlink, doesn't it?

Author: There's nothing more to add to the contributing docs about this, is there?

Author: @geekosaur are you happy to resolve this thread?

Reviewer: I'd prefer a warning about having to switch `fourmolu` versions, possibly with a reference to #9230, but otherwise it's fine.

Reviewer: Well, except [we already have problems with `--program-suffix`](https://github.com/haskell/cabal/issues/10290) and I really wouldn't be surprised if they also extend to `--program-prefix` as well. (Although, isn't that suffix instead of prefix?)

Author: @geekosaur, I added a warning about switching fourmolu versions and referenced #9230.

> ```

> [!Tip]

> > If you have multiple versions of GHC installed with `ghcup`, a series specific

Reviewer 2: ```suggestion
> If you have multiple versions of GHC installed with `ghcup`, a series-specific
```
Hyphens strikes again...

> ghc can be used by adding an option of `--with-compiler=ghc-x.y` (for the
> latest installed `x.y.z` version in the `x.y` series) or an option of
> `--with-compiler=ghc-x.y.z` to the above install command.
>
> ```
> $ ghc --version | sed 's/[[:alpha:]|(|,|[:space:]]//g'
> 9.10.1
>
> $ ghc-9.6 --version | sed 's/[[:alpha:]|(|,|[:space:]]//g'
> 9.6.6
>
> $ ghc-9.4.8 --version | sed 's/[[:alpha:]|(|,|[:space:]]//g'

> > 9.4.8

Reviewer 3: Isn't this equivalent to `ghc --numeric-version`?

Reviewer: It is, in versions that have it… but I'm pretty sure we don't support ghc 7.0 or earlier any more.

Author: Thanks for the tip @jasagredo. I'm using `--numeric-version` now.
